runtime/jvm/src/test/kotlin/kotlinx/serialization/formats/json/JsonElementReaderWriterTest.kt

```diff
@@ -0,0 +1,222 @@
+            assertEquals(1000, timestamp)
+        }
+    }
+
+    @Test
+    fun testParseErrorString() {
+        val ev = JSON.parse(Event.serializer(), inputErrorString)
+        with(ev) {
+            assertEquals(1, id)
+            assertEquals(DummyEither.Left("Connection timed out"), payload)
+            assertEquals(1001, timestamp)
+        }
+    }
+
+    @Test
+    fun testWriteDataString() {
+        val outputData = Event(0, DummyEither.Right(Payload(42, 43, "Hello world")), 1000)
+        val ev = JSON.stringify(Event.serializer(), outputData)
+        assertEquals(inputDataString, ev)
+    }
+
+    @Test
+    fun testWriteDataStringUnquoted() {
+        val outputData = Event(0, DummyEither.Right(Payload(42, 43, "Hello world")), 1000)
+        val ev = JSON.unquoted.stringify(Event.serializer(), outputData)
+        assertEquals("""{id:0,payload:{msg:"Hello world",from:42,to:43},timestamp:1000}""", ev)
+    }
+
+    @Test
+    fun testWriteDataStringIndented() {
+        val outputData = Event(0, DummyEither.Right(Payload(42, 43, "Hello world")), 1000)
+        val ev = JSON.indented.stringify(Event.serializer(), outputData)
+        assertEquals("""{
+            |    "id": 0,
+            |    "payload": {
+            |        "msg": "Hello world",
+            |        "from": 42,
+            |        "to": 43
+            |    },
+            |    "timestamp": 1000
+            |}""".trimMargin(), ev)
+    }
+
+    @Test
+    fun testWriteErrorString() {
+        val outputError = Event(1, DummyEither.Left("Connection timed out"), 1001)
+        val ev = JSON.stringify(Event.serializer(), outputError)
+        assertEquals(inputErrorString, ev)
+    }
+
+    @Test
+    fun testParseDataJson() {
+        val ev = JsonTreeMapper().readTree(inputDataJson, Event.serializer())
+        with(ev) {
+            assertEquals(0, id)
+            assertEquals(DummyEither.Right(Payload(42, 43, "Hello world")), payload)
+            assertEquals(1000, timestamp)
+        }
+    }
+
+    @Test
+    fun testParseErrorJson() {
+        val ev = JsonTreeMapper().readTree(inputErrorJson, Event.serializer())
+        with(ev) {
+            assertEquals(1, id)
+            assertEquals(DummyEither.Left("Connection timed out"), payload)
+            assertEquals(1001, timestamp)
+        }
+    }
+
+    @Test
+    fun testWriteDataJson() {
+        val outputData = Event(0, DummyEither.Right(Payload(42, 43, "Hello world")), 1000)
+        val ev = JsonTreeMapper().writeTree(outputData, Event.serializer())
+        // JsonObject#equals seems to be broken here.
+        assertEquals(inputDataJson.toString(), ev.toString())
+    }
+
+    @Test
+    fun testWriteErrorJson() {
+        val outputError = Event(1, DummyEither.Left("Connection timed out"), 1001)
+        val ev = JsonTreeMapper().writeTree(outputError, Event.serializer())
+        // JsonObject#equals seems to be broken here.
```

Comment:
I've fixed this in https://github.com/Kotlin/kotlinx.serialization/commit/6d4016bd8b2f45c3d2167f16055c6cdb59ea28ad, so you probably should reopen this PR against dev.

Reply:
Thanks, will rebase on current `dev` and get rid of this hack.